### ISSUE TITLE
strconv.ParseInt invalid syntax

### ISSUE THREAD

**Issue by scalp42** [hashicorp/terraform#5254]
Using Terraform v0.6.12-dev (e70516a286493cf6e9b1a8a5649c784d89684fa5), we're now getting a syntax error:

```
terraform apply terraform
Error applying plan:

1 error(s) occurred:

* Error reading aws_subnet.infra-public count: strconv.ParseInt: parsing "${var.infra_count}": invalid syntax
```

This is the config:

```go
variable "infra_count" {
  description = "Count index for all the things"
  default = 3
}

resource "aws_subnet" "infra-public" {
  count = "${var.infra_count}"
  vpc_id                  = "${aws_vpc.infra.id}"
  availability_zone       = "${concat(var.region, lookup(var.infra_subnet_availability_zones, count.index))}"
  cidr_block              = "${lookup(var.infra_public_subnet_cidr_blocks, count.index)}"
  map_public_ip_on_launch = true
  depends_on              = ["aws_internet_gateway.infra-igw"]
  tags {
    Name = "infra-public-${concat(var.region, lookup(var.infra_subnet_availability_zones, count.index))}"
    Description = "infra-public-${concat(var.region, lookup(var.infra_subnet_availability_zones, count.index))}"
    vpc = "infra"
    terraform = "true"
  }
}
```

Git blame:

```
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 11) resource "aws_subnet" "infra-public" {
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 12)   count = "${var.infra_count}"
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 13)   vpc_id                  = "${aws_vpc.infra.id}"
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 14)   availability_zone       = "${concat(var.region, lookup(var.infra_subnet_availability_zones, count.index))}"
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 15)   cidr_block              = "${lookup(var.infra_public_subnet_cidr_blocks, count.index)}"
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 16)   map_public_ip_on_launch = true
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 17)   depends_on              = ["aws_internet_gateway.infra-igw"]
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 18)   tags {
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 19)     Name = "infra-public-${concat(var.region, lookup(var.infra_subnet_availability_zones, count.index))}"
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 20)     Description = "infra-public-${concat(var.region, lookup(var.infra_subnet_availability_zones, count.index))}"
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 21)     vpc = "infra"
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 22)     terraform = "true"
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 23)   }
d71c2a2c (Anthony Scalisi 2015-07-02 22:49:36 -0700 24) }


8261d4fb (Anthony Scalisi 2015-07-02 22:49:20 -0700  1) variable "infra_count" {
8261d4fb (Anthony Scalisi 2015-07-02 22:49:20 -0700  2)   description = "Count index for all the things"
8261d4fb (Anthony Scalisi 2015-07-02 22:49:20 -0700  3)   default = 3
8261d4fb (Anthony Scalisi 2015-07-02 22:49:20 -0700  4) }
```

Error was not happening up to https://github.com/hashicorp/terraform/commit/5748487af943a8b4cc16e2966f33c5c54d50333d

Any ideas ?

**Comment by phinze** [comment]
Hi @scalp42 - this is definitely odd. It might also have to do with https://github.com/hashicorp/terraform/pull/4795 which has more of a diff in the area and was merged around the same time.

I'm trying to reproduce on this side and not having any luck. Can you spot any differences between the config below and yours that might be key to reproducing the issue?

Also: does the syntax error happen every time or only sporadically?

If you could [gist](https://gist.github.com) up your [debug log](https://www.terraform.io/docs/internals/debugging.html) (scrubbed of any sensitive information) when you hit the error that would be helpful as well.

```tf
variable "region" {
  default = "us-west-2"
}

variable "infra_count" {
  default = 3
}

variable "infra_public_subnet_cidr_blocks" {
  default = {
    "0" = "10.0.1.0/24"
    "1" = "10.0.2.0/24"
    "2" = "10.0.3.0/24"
  }
}

variable "infra_subnet_availability_zones" {
  default = {
    "0" = "a"
    "1" = "b"
    "2" = "c"
  }
}

resource "aws_vpc" "infra" {
  cidr_block = "10.0.0.0/16"
}

resource "aws_internet_gateway" "infra-igw" {
  vpc_id = "${aws_vpc.infra.id}"
}

resource "aws_subnet" "infra-public" {
  count                   = "${var.infra_count}"
  vpc_id                  = "${aws_vpc.infra.id}"
  availability_zone       = "${concat(var.region, lookup(var.infra_subnet_availability_zones, count.index))}"
  cidr_block              = "${lookup(var.infra_public_subnet_cidr_blocks, count.index)}"
  map_public_ip_on_launch = true
  depends_on              = ["aws_internet_gateway.infra-igw"]
  tags {
    Name = "infra-public-${concat(var.region, lookup(var.infra_subnet_availability_zones, count.index))}"
    Description = "infra-public-${concat(var.region, lookup(var.infra_subnet_availability_zones, count.index))}"
    vpc = "infra"
    terraform = "true"
  }
}
```

**Comment by scalp42** [comment]
Thanks for the quick answer @phinze 

As of right now I can reproduce it all the time on any `terraform apply` (so blocked):

This ![gist](https://gist.github.com/scalp42/3b3bb51f4e37fbaf2434) contains the steps that got me there:

- set desired capacity to 1 in AWS by hand
- bump desired capacity from 0 to 2 in terraform config (explains I think why it's weird re diff)
- ran terraform plan
- ran terraform apply which lead to error re syntax

I'll send you the trace by email as it's really big in a few minutes

**Comment by phinze** [comment]
(Got the log in email, will analyze to help cook up a repro case.)

Noting here that you can workaround the issue by temporarily inlining the value of `infra_count`, which should unblock you.

**Comment by phinze** [comment]
Okay got a repro here.

### Config

```tf
variable "infra_count" {
  default = 3
}

resource "aws_vpc" "infra" {
  cidr_block = "10.0.0.0/16"
}

resource "aws_internet_gateway" "infra-igw" {
  vpc_id = "${aws_vpc.infra.id}"
}

resource "aws_subnet" "infra-public" {
  count                   = "${var.infra_count}"
  vpc_id                  = "${aws_vpc.infra.id}"
  availability_zone       = "us-west-2a"
  cidr_block              = "10.0.${count.index+1}.0/24"
  map_public_ip_on_launch = true
  depends_on              = ["aws_internet_gateway.infra-igw"]
}

resource "aws_launch_configuration" "foo" {
  image_id      = "ami-03de3c63"
  instance_type = "t2.nano"
}

resource "aws_autoscaling_group" "foo" {
  vpc_zone_identifier  = ["${aws_subnet.infra-public.*.id}"]
  launch_configuration = "${aws_launch_configuration.foo.id}"
  min_size             = 0
  max_size             = 0  # Change me from 0 -> 1 after first apply
}
```

### Steps

1. Apply config as-is w/ plain `terraform apply`
1. Change `aws_autoscaling_group.foo.max_size` from `0` to `1`
1. Generate a planfile `terraform plan -out maxsize.tfplan`
1. Apply the planfile `terraform apply maxsize.tfplan`

### Expected behavior:

Plan is executed as displayed:

```
~ aws_autoscaling_group.foo
    max_size: "0" => "1"


Plan: 0 to add, 1 to change, 0 to destroy.
```

### Observed behavior:

```
Error applying plan:

1 error(s) occurred:

* Error reading aws_subnet.infra-public count: strconv.ParseInt: parsing "${var.infra_count}": invalid syntax

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

### Terraform version

Current master: b831ba3ab6fe527cbd59229a266fa8fcf65ec5a8

**Comment by phinze** [comment]
Tighter repro w/ `template_file`:

```tf
variable "c" { default = 1 }
resource "template_file" "parent" {
  count = "${var.c}"
  template = "Hi"
}
resource "template_file" "child" {
  template = "${join(",", template_file.parent.*.rendered)} ok"
}
```

```bash
terraform apply # creates resources
# modify template_file.child.template to toggle add/remove extra "ok" string
terraform plan -out template.tfplan
terraform apply template.tfplan
# ...
* Error reading template_file.parent count: strconv.ParseInt: parsing "${var.c}": invalid syntax
```

**Comment by scalp42** [comment]
:bow: @phinze thanks a lot for looking into it

**Comment by pmoust** [comment]
Yeap, this breaks our TF-managed when using HEAD. 

A member of my team reported this happening with interpolation functions + variables in 0.6.9 as well, this happens when a var is referenced from a resource to another >1 levels deep. will try and get a small repro case from him

/cc @alexsapran

**Comment by phinze** [comment]
`git bisect`ed introducing commit as https://github.com/hashicorp/terraform/pull/5026

Root cause of OP's issue is uninterpolated var making it to [here](https://github.com/hashicorp/terraform/blob/e1e39eefbc94486a702737849616770afc51ddfc/terraform/interpolate.go#L432-L438), but only when using a plan file.

Either there's a mistake in #5026 or it uncovered a latent dependency here on a side-effect of interpolation elsewhere. Either way, will get this sorted out before release!

**Comment by Gary-Armstrong** [comment]
I just wanted to throw in my error which is essentially the same, although mine is not reliably reproducible.  This will happen seemingly randomly, but will succeed (or produce same error on a different count var) immediately on the next apply.

`Error configuring: 1 error(s) occurred:

* strconv.ParseInt: parsing "${var.head_count}": invalid syntax`

**Comment by scalp42** [comment]
Going to try master, thanks a lot for looking into it @mitchellh @phinze

**Comment by davedash** [comment]
I'm running into a similar issue today, not sure if I need to create a new ticket.

```
* module.prod.module.esw_instance.output.deps: Error reading aws_route53_record.external count: strconv.ParseInt: parsing "${aws_instance.instance.count}": invalid syntax
```

Previously the count value was `${var.enabled?1:0}`.  This is only a problem when I build a module that depends on the module named `prod`.

I'm posting first to see if anybody still has similar issues and if this is new, I can try to provide more detail.  I'm just not sure what's helpful.
